azurerm/internal/services/network/firewall_policy_rule_collection_group_resource.go

```diff
@@ -0,0 +1,816 @@
+										Elem: &schema.Schema{
+											Type:         schema.TypeString,
+											ValidateFunc: validate.FirewallPolicyRulePort,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"nat_rule_collection": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MinItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"priority": {
+							Type:         schema.TypeInt,
+							Required:     true,
+							ValidateFunc: validation.IntBetween(100, 65000),
+						},
+						"action": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								// Hardcode to using `Dnat` instead of the one defined in Swagger (i.e. network.DNAT) because of: https://github.com/Azure/azure-rest-api-specs/issues/9986
+								// Setting `StateFunc: state.IgnoreCase` will cause other issues, as tracked by: https://github.com/hashicorp/terraform-plugin-sdk/issues/485
+								// Another solution is to customize the hash function for the containing block, but as there are a couple of properties here, especially
+								// has property whose type is another nested block (Set), so the implementation is nontrivial and error-prone.
+								"Dnat",
+							}, false),
+						},
+						"rule": {
+							Type:     schema.TypeSet,
+							Required: true,
+							MinItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validate.FirewallPolicyRuleName(),
+									},
+									"protocols": {
+										Type:     schema.TypeSet,
+										Required: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+											ValidateFunc: validation.StringInSlice([]string{
+												string(network.FirewallPolicyRuleNetworkProtocolTCP),
+												string(network.FirewallPolicyRuleNetworkProtocolUDP),
+											}, false),
+										},
+									},
+									"source_addresses": {
+										Type:     schema.TypeSet,
+										Optional: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
+									},
+									"source_ip_groups": {
+										Type:     schema.TypeSet,
+										Optional: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
+									},
+									"destination_address": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validation.IsIPAddress,
+									},
+									"destination_ports": {
+										Type:     schema.TypeSet,
+										Optional: true,
+										Elem: &schema.Schema{
+											Type:         schema.TypeString,
+											ValidateFunc: validate.FirewallPolicyRulePort,
```

Comment:
could we ue the plugin SDK validater here?

Reply:
In fact, the API used here allows not only a simple port number, it also supports a port range (e.g. `1000-2000`). Hence, we need to extend the SDK port validation function a bit here.